File: letni25-26/aisd/pracownie/pracownia5/E.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <stack>
#include <cstdio>

using namespace std;

vector<pair<int,int>> DIRS = {{0,1},{0,-1},{1,0},{-1,0}};


int find_time(int x, vector<int>& xs) {
    int l = 0;
    int r = xs.size();

    while (l < r)
    {
        int m = (l + r) / 2;

        if (xs[m] < x) {
            l = m + 1;
        }
        else {
            r = m;
        }
    }

    return l - 1;
    
}

class UnionFind {
    private:
        vector<int> parent;
        vector<int> size;
        
    public:
        UnionFind(int n) {
            parent.resize(n);
            size.assign(n, 1);
            for (int i = 0; i < n; i++) {
                parent[i] = i;
            }
        }

        int find(int p) {
            if (p == parent[p]) {
                return p;
            }

            parent[p] = find(parent[p]);
            return parent[p];

        }

        int union_disj(int a, int b) {
            int x = find(a);
            int y = find(b);

            if (x == y) {
                return 0;
            }

            if (size[x] > size[y]) {
                swap(x,y);
            }

            parent[y] = x;
            size[x] += size[y];

            return 1;
        }
};

    
vector<int> solve(int n,int m,vector<vector<int>>& board,vector<int>& times) {
    unordered_map<int, vector<pair<int,int>>> pos_for_time;
    
    for(int i = 0; i < m; i ++) {
        for(int j = 0; j < n; j++) {
            int t = find_time(board[i][j], times);
            if (t == -1) {
                continue;
            }
            pos_for_time[times[t]].push_back({i,j});
        }
    }

    reverse(times.begin(), times.end());

    int prev = -1;
    int islands_prev = 0;
    vector<int> island_counts(times.size());


    UnionFind uf = UnionFind(n*m);

    for(auto t : times) {
        if (t == prev) {
            island_counts.push_back(island_counts.back());
            continue;
        }

        vector<pair<int,int>> elems = pos_for_time[t];

        int count_unions = 0;

        for(auto [y,x] : elems) {
            for (auto [dx, dy] : DIRS) {
                int xp = x + dx;
                int yp = y + dy;

                if (0 <= xp && xp < n && 0 <= yp && yp < m) {
                    if (board[yp][xp] > t) {
                        int r = uf.union_disj(n*y+x,n*yp+xp); // r is 0 or 1
                        count_unions += r;
                    }
                }
            }
        }

        int islands_now = islands_prev + elems.size() - count_unions;
        island_counts.push_back(islands_now);
        islands_prev = islands_now;
        prev = t;
    }

    reverse(island_counts.begin(), island_counts.end());

    return island_counts;


}





int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int x;
    int y;
    

    cin >> y >> x;

    vector<vector<int>> board(y,vector<int>(x));

    for(int i = 0; i < y; i ++) {
        for(int j = 0; j < x; j ++) {
            int h;
            cin >> h;
            board[i][j] = h;
        }
    }

    int T;
    cin >> T;

    vector<int> times(T);

    for(int i = 0; i < T; i ++) {
        int time;
        cin >> time;
        times[i] = time;
    }

    vector<int> island_counts = solve(x,y,board,times);


    for(int i = 0; i < T; i++) {
        cout << island_counts[i] << ' ';
    }
    cout << endl;
}

File: letni25-26/aisd/pracownie/pracownia5/E.py
DIRS = [(0,1),(0,-1),(1,0),(-1,0)]


def find_time(x,xs):
    pass
    #xs posortowane, chcemy najwiekszei  t ze xs[i] < x
    # l = 0
    # r = len(xs)-1
    # while l < r:
    #     m = (l+r)//2
    #     print(x, m, xs[m])
    #     time.sleep(0.1)
    #     if x > xs[m]:
    #         l = m + 1
    #     elif x == xs[m]:
    #         return m-1 if m != 0 else -1
    #     elif x < xs[m]:
    #         r = m

    # return l-1
def find_time(x, xs):
    l = 0
    r = len(xs)  # ważne: półotwarty przedział [l, r)

    while l < r:
        m = (l + r) // 2
        if xs[m] < x:
            l = m + 1
        else:
            r = m

    return l - 1

# print('dla : ',0,find_time(0,[1,2,3,4,5]))
# print('dla : ',1,find_time(1,[1,2,3,4,5]))
# print('dla : ',2,find_time(2,[1,2,3,4,5]))
# print('dla : ',3,find_time(3,[1,2,3,4,5]))
# print('dla : ',4,find_time(4,[1,2,3,4,5]))
# print('dla : ',5,find_time(5,[1,2,3,4,5]))
# print('dla : ',6,find_time(6,[1,2,3,4,5]))
class UnionFind:
    def __init__(self,xs):
        self.parents = {x : x for x in xs}
        self.size = {x : 1 for x in xs}
        
        #kazdy ma wskaznik na rodzica
        #reprezentujemy 'drzewo'

    def find(self,x):
        if x == self.parents[x]:
            return x
        
        #kompresja sciezki
        self.parents[x] = self.find(self.parents[x])
        return self.parents[x]

    def union(self,a,b):
        #returns if a union was succesful
        x = self.find(a)
        y = self.find(b)

        if x == y:
            return 0 
        
        
        #poz wieksze podpinamy mniejsze
        if self.size[x] < self.size[y]:
            x,y = y,x

        self.parents[y] = x
        self.size[x] += self.size[y]

        return 1

    


def solve(n,m,board,times):
    #chcemy tu miec wszystkie komorki ktore nie beda zalane w dniu t i zalane w czasie t+1
    print(f'n m to : {n} {m}')
    #moga byc te same wartosci w times
    pos_for_time = {t : [] for t in times}
    

    for i in range(m):
        for j in range(n):
            t = find_time(board[i][j],times)
            if t == -1:
                continue
            pos_for_time[times[t]].append((i,j))

    print(pos_for_time)
    rev_times = times[::-1]
    print(rev_times)
    prev = None
    islands_prev = 0
    counts = []

    uf = UnionFind([(j,i) for i in range(n) for j in range(m)])

    for t in rev_times:
        
        if t == prev:
            counts.append(counts[-1])
            continue

        
        elems = pos_for_time[t]


        count_unions = 0

        for y,x in elems:
            neighbours = [(x+dx,y+dy) for dx,dy in DIRS if 0 <= x+dx and x + dx < n and 0 <= y+dy and y+dy < m]
            
            
            for xp,yp in neighbours:
                
                # print(x,y, xp,yp)
                if board[yp][xp] > t: #nie jest zatopiony
                    # print(t,board[yp][xp])
                    r = uf.union((y,x),(yp,xp)) #0 or 1
                    count_unions += r

        islands_now = islands_prev + len(elems) - count_unions   
        counts.append(islands_now)
        islands_prev = islands_now
        prev = t

    return counts[::-1]
    #zaczynamy od ostatniego i robimy union find-a


with open('inp3.txt','r') as inp:
    data = inp.read().splitlines()
    y,x = data[0].split()
    x,y = int(x),int(y)
    board = [[int(i) for i in r.split()] for r in data[1:1+y]]

    times = [int(i) for i in data[-1].split()]
    print(data)
    print(x,y)
    print(board)
    print(times)
    islands = solve(x,y,board,times)
    print(islands)